### ISSUE TITLE
Question about in icl_lm_eval in evaluate.py

### ISSUE THREAD

**Issue by asimokby** [zjunlp/EasyEdit#426]
In evaluate.py, in the icl_lm_eval fucntion, the following line is in the else block. 

`
ans = torch.argmax(logits, dim=-1)[:,-target_ids.size(1):-1].squeeze()
`

Why do we slice columns from -target_ids.size(1)? Doesn't this ignore tokens at the start? For example, if target_ids = 10, and the output is 15 tokens, then we neglect the first 4 tokens, right? Why do you cutoff the first 4 tokens in this case and not the last 4?

Thanks!

**Comment by asimokby** [comment]
Thanks a lot. I understand. 

My question now is why do we give the target, which is the answer, in the prompt itself? What are we really testing in this case? I feel like I am missing something. Could he answer please? 

```python
encodings = tokenizer(f'{x} {target}', return_tensors='pt')
```

Here is a complete example: 

```python
tokenizer = GPT2Tokenizer.from_pretrained('asimokby/lte-gpt2')
model = GPT2LMHeadModel.from_pretrained('asimokby/lte-gpt2')

x = """Please acknowledge the updated information provided below and respond to the subsequent query.

[Updated Information]:
What day was USA-199 launched? 20 December 2007

[Query]:
When was GPS IIR-18(M) launched?"""

target = "20 December 2007"

target_ids = tokenizer(' ' + target, return_tensors='pt')['input_ids']
encodings = tokenizer(f'{x} {target}', return_tensors='pt')
input_ids = encodings['input_ids']
logits = model(input_ids=input_ids, attention_mask=encodings['attention_mask']).logits
ans = torch.argmax(logits, dim=-1)[::].squeeze()
ans = torch.argmax(logits, dim=-1)[:,-target_ids.size(1)-1:-1].squeeze()
tokenizer.decode(ans)
```

Output: 
```]
 20 December 2007
```

**Comment by JizhanFang** [comment]
First, as an ICL (In-Context Learning) editing method, it's quite straightforward to provide the answer to a question in the prompt, right? Just like the `20 December 2007` in the `[Updated Information]` part of your code.
I guess your confusion lies in why we still need to append the target to the end of the prompt, as shown in this line of code:
```
encodings = tokenizer(f'{x} {target}', return_tensors='pt')
```
In fact, this is exactly the process of running `model()` that I explained earlier. Although we append the target to the prompt, this does not actually lead to label leakage. We are still making the model give its prediction based on the prompt: assuming there are 50 tokens totally in your x (I haven't counted them, but the exact number doesn't matter), when the model predicts the 51st token, i.e., the token right after `launched?`, it can only see the first 50 tokens. At this point, the 51st token, which corresponds to the "20" in the target, is not visible to the model during this prediction step. Therefore, the model is still making predictions based on the prompt, which explains why our evaluation method is valid. I hope this clarifies your confusion.

**Comment by asimokby** [comment]
My bad. I mixed model.generate() with model(). All clear now. Thanks a lot.
